tests/filter-video-average.cpp

```diff
@@ -0,0 +1,189 @@
+    OK(acquire_start(runtime));
+    {
+        uint64_t nframes = 0;
+        const uint64_t expected_nframes = props.video[0].max_frame_count / props.video[0].frame_average_count;
+        LOG("Expecting %d frames", expected_nframes);
+
+        // Each pixel is drawn from a uniform distribution in [0, 255].
+        // Without averaging we would expect the within-frame pixel value
+        // variance to follow that of a discrete uniform distribution:
+        // (256^2 - 1) / 12.
+        // By averaging over every two frames, this shrinks by a factor of 2:
+        // (256^2 - 1) / 24
+        const double expected_pixel_variance = 2730.625;  // (256*256 - 1) / 24
+        const size_t num_pixels = props.video[0].camera.settings.shape.x * props.video[0].camera.settings.shape.y;
+        const double normalization_factor = 1.0 / (num_pixels * expected_nframes);
+        double actual_pixel_mean = 0;
+        double actual_pixel_sum_of_squares = 0;
+
+        while (nframes < expected_nframes) {
+            struct clock throttle
+            {};
+            clock_init(&throttle);
+            EXPECT(clock_cmp_now(&clock) < 0,
+                   "Timeout at %f ms",
+                   clock_toc_ms(&clock) + time_limit_ms);
+            VideoFrame *beg, *end, *cur;
+            OK(acquire_map_read(runtime, 0, &beg, &end));
+            for (cur = beg; cur < end; cur = next(cur)) {
+                LOG("stream %d counting frame w id %d", 0, cur->frame_id);
+                CHECK(cur->shape.dims.width ==
+                      props.video[0].camera.settings.shape.x);
+                CHECK(cur->shape.dims.height ==
+                      props.video[0].camera.settings.shape.y);
+                float * data = (float *)&(cur->data[0]);
+                for (size_t i=0; i < num_pixels; ++i) {
+                    const double value = (double)data[i];
+                    actual_pixel_mean += normalization_factor * value;
+                    actual_pixel_sum_of_squares += normalization_factor * value * value;
+                }
+                ++nframes;
+            }
+            {
+                uint32_t n = (uint32_t)consumed_bytes(beg, end);
+                OK(acquire_unmap_read(runtime, 0, n));
+                if (n)
+                    LOG("stream %d consumed bytes %d", 0, n);
+            }
+            clock_sleep_ms(&throttle, 100.0f);
+
+            LOG("stream %d nframes %d. remaining time %f s",
+                0,
+                nframes,
+                -1e-3 * clock_toc_ms(&clock));
+        }
+
+        CHECK(nframes == expected_nframes);
+        // Our tolerance is a little loose since the pixel values are high
+        // and we're only averaging over every two frames.
+        double actual_pixel_variance = actual_pixel_sum_of_squares - (actual_pixel_mean * actual_pixel_mean);
+        LOGE("pixel variance: actual = %g, expected = %g", actual_pixel_variance, expected_pixel_variance);
```

Comment:
Every time I run this, Windows or Linux, I get 2730.62. Same for you?

Reply:
On macOS, I get 2729.74 every time

The RNG is seeded, though after a quick look at the implementation, I'm a little surprised there are cross-platform differences.

Reply:
Yeah, I mentioned this to Nathan and he clued me in to the seed. I also see 2729.4 on the Mac. That is pretty weird.

Reply:
I'm not sure how close I'd expect to get to the true variance (based on the number of samples), but if we really feel strongly, I can do the math!